expo/expo
# [tools] Switch expo home publishing to EAS update

# Why

This is an effort to switch dev home publishing (the feature that uses a published update so that running Expo Go locally doesn't require building the bundle/manifest at the same time and can instead load it remotely) to EAS Update.

Currently it uses classic updates, which are being discontinued.

# How

1. Switch `et pdh` to EAS update. This puts an EAS Update group URL in `dev-home-config.json`
2. Update macros (constants generated during native build of Expo Go) to support new manifests. Embed the asset request headers necessary to access bundles on the CDN into the constants as well.
3. Update the native code that reads the constants to read asset request headers as well.
4. Update android to include the asset request header in the bundle asset request.
5. Update iOS to use expo-updates to download the home bundle. This required changing the visibility of a bunch of stuff in the library. (couldn't do the same change as android since the utility we were using to download the home bundle in the past doesn't support custom headers and lives in react-native). This has the secondary benefit of us being able to remove a lot of Expo Go code from iOS in a subsequent PR. We may also want to make all visibility stuff in the expo-updates library consistent rather than just what we need.

Note: I haven't yet published to prod project, just staging so far. Will publish to prod project upon approval before landing and update dev-home-config to a prod URL.

# Known Issues

- This doesn't handle normal kernel publishing (running `expo publish` within the home directory) which is part of the SDK release process. This will need to be done in a subsequent PR. See code like: https://github.com/expo/expo-cli/blob/af2874cf685e6562c74ab597cdb099750c84a3fd/packages/xdl/src/project/publishAsync.ts#L252
- There are lots of other apps in the `apps` directory that still use classic publishing using the `et publish-app` command. These will also need to be switched over.
- Pretty much any tools that rely upon `expo publish` will need to be switched.

# Test Plan

1. Swap out project IDs for one in staging
7. `et pdh`
8. See it publishes and updates the URL in `dev-home-config.json` to an EAS update group URL.
9. Run android and ios and ensure the EAS update manifest is downloaded in the macros.
10. Switch to running home bundler locally, check that setting LOCAL works in iOS and in android the local manifest is still picked up automatically.

# Checklist
